python/ray/data/datasource/tfrecords_datasource.py

```diff
@@ -64,13 +64,10 @@ def _write_block(
 
 def _convert_example_to_dict(
     example: "tf.train.Example",
-) -> Dict[str, Union[bytes, List[bytes], float, List[float], int, List[int]]]:
+) -> Dict[str, Union[List[List[bytes]], List[List[float]], List[List[int]]]]:
     record = {}
     for feature_name, feature in example.features.feature.items():
-        value = _get_feature_value(feature)
-        if len(value) == 1:
-            value = value[0]
```

Comment:
I don't know why previously we have this special handling for the list with single element. Is it okay to remove? cc @bveeramani and @clarkzinzow.

Reply:
There's no way to create scalars with TFRecords. If your features represent scalars (as is usually the case), you still have to use list like `FloatList`.

Let's say you're working with the Iris dataset. Your TFRecords would look like this:

```
feature={
            "sepal.height": tf.train.Feature(float_list=tf.train.FloatList(value=[3.5])),
            "sepal.width": tf.train.Feature(float_list=tf.train.FloatList(value=[3.2])),
            "species": tf.train.Feature(bytes_list=tf.train.BytesList(value=[b"iris"])),
        }
```

If we didn't special case, your dataset that looks like:

```
Dataset(..., schema={"sepal.height": object, "sepal.width": object, "species": object}
```

This makes it impossible to use preprocessors and difficult to write UDFs.

What you probably want (which is what you get with the special casing) is a dataset like this:

```
Dataset(..., schema={"sepal.height": float, "sepal.width": float, "species": object}
```

I understand that special casing is generally bad, but based on the examples I've seen online, it seems like single-elements list represent scalars in the vast majority of cases. I've never seen a use case where a single-element list actually represents a list.

If we remove the special casing, `read_tfrecords` might not work the way users expect it to.

Reply:
@bveeramani - thanks for context. It makes sense to me, changed to maintain existing behavior.